### ISSUE TITLE
[question] Should profile [env] vars affect header only package hashes?

### ISSUE THREAD

**Issue by aleksa2808** [conan-io/conan#8662]
Hi,
while packaging a header only library (marked with `header_only()`), I noticed I got a different resulting package hash for different profiles I used. It turned out that the metadata bundled with the package (`conaninfo.txt`) contained the `[env]` fields of each specific profile, which in my case were the CC and CXX variables. Since I use the `package_revision_mode` for package_id resolution this resulted in my parallel CI job branches (one for each profile) stepping over each other since each one used their own package revision of this header lib for its downstream dependency package_id resolution, yet when they all upload their packages only the last revision of the header only lib remains usable (since they all have the same package_id). This results in the libs downstream dependency packages built with any other profile except the last one to upload needing an immediate rebuild.

I tried to picture the described scenario with the following diagram:

![Untitled Diagram](https://user-images.githubusercontent.com/7463511/111495276-7cb72980-873f-11eb-8845-72d87b9628dd.png)

On this picture, the package of the header-only lib1 with with the revision `zzz` is uploaded last and overwrites the lib1 package with revision `yyy` since they share a package_id. But, lib2 on the left was already built using lib1 with package revision `yyy`, so the lib1 overwrite would render that version of lib2 outdated as well since now it requires a build with lib1 package revision `zzz`.

Because of the described issues I was wondering if it's necessary to include the profile `[env]` fields into the metadata of a package marked with `header_only()`?

Conan version: 1.33.1

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by jgsogo** [comment]
Hi, @aleksa2808 . Really interesting scenario the one you share, and it is something we need to think about. There is a lot of work related to environment variables, configuration and package-ids for the next Conan 2.0 version.

(Maybe workaround) In ConanCenter, for example, we compute the package-ids for all the configurations, then we remove duplicates and finally we build only those different (for header-only libraries we trigger only one configuration). That would prevent your problem, but it requires some extra orchestration in your CI.

Most of the time if those variables are there it is because they affect the generated package: Conan cannot know if the recipe contains some `if/else` based on any environment variable, or maybe the `.h` files are generated taking into account those values.

The safest for Conan is to assume that environment variables (if they are in the profile, we are not taking into account all the environment) modify the package and, then, they should be added to the package hash.

BUT, let's add this to the backlog for Conan 2.0. I think it is something to consider and decide.

**Comment by aleksa2808** [comment]
I understand that the environment variables may modify the output of the package and this is something that they share with settings and options, since someone can do an `if/else` to generate different headers based on OS, detected compiler etc. That's why I assumed that the env vars fall into the same bucket and that the `header_only()` directive would eliminate their respective metadata as well. For scenarios I have and the ones I imagine, I think making `header_only()` remove the `[env]` fields as well would do the trick, but I'm not sure if there are some situations where the current behavior is more desired. If modifying the behavior of this method would be considered a breaking change, another approach might be to have some different way/method to specify that the environment variables do not affect the package output.

I know that the environment vars are a tricky topic for Conan and definitely much less controlled than settings and options, so I appreciate the effort you're making on them for 2.0. Are there some specific issues on this topic that I can take a look at?

As for the CCI workaround, my current approach is to build each profile in parallel so making one profile build some packages used in other profiles seems complicated since it would involve some kind of barrier logic between the parallel jobs. How do you deal with this problem? Would the lockfile build-order functionality be helpful here?

**Comment by solvingj** [comment]
> How do you deal with this problem? Would the lockfile build-order functionality be helpful here?
@aleksa2808  I'm glad you asked this question now. 

The new feature of "Lock Bundles" was created to deal with two sources of complexity when using Conan within CI pipelines.  Indeed, one of those sources was this one.  Whenever CI pipelines rebuild the same package ID multiple times, it can have several negative side effects:
- Inefficient, waste of resources
- Different recipe revisions (rrev) due to line endings
- Upload conflicts
- With lockfiles, updated lockfiles each point to different package revisions (prev)

Of note, there are still some valid use-cases and reasons to build the same package ID multiple times in a pipeline, so CI pipeline authors need the ability to choose. 

In short, we acknowledged a few months ago that it was simply too difficult and complex for users to identify redundant package ID builds, and design CI pipelines on a per-package-id basis rather than per-profile basis or per-lockfile basis.  

The lockfile bundle feature is our first contribution toward making this goal easier. However, it's tightly coupled to the lockfiles feature, so if you're still doing CI pipelines based on profiles, it is of no help there.  Since most pipelines are currently profile-based, now is probably a good time to look at what might be possible for that case.  

Ping @memsharded . First thought is something like `conan profile bundle`.  Obviously, there's no concept of build order like we have in lockfiles, but there's a concept of de-duplicating package_ids across N lockfiles.  It seems possible to take a list of profiles, pre-calculate the package ID's, and then produce some json output with the mapping similar to lock bundle. Then, users could pick `first` profile when appropriate.  I dunno, maybe it's the wrong abstraction, just was my first thought to build off what we have and stay consistent.

**Comment by aleksa2808** [comment]
If the Conan development effort will start (or already is) preferring the per-package-id CI pipelines, that would certainly be an answer of sorts to the original issue, since it solves the underlying problem that I had. Although, it still feels a bit weird seeing a `CC=/usr/bin/gcc-9` in the metadata of a header only package, especially since I assume that defining the `CC`/`CXX` variables in profiles seems like a pretty standard thing to do.

As for the `conan profile bundle` suggestion, this does sound like calculating a "profile build order". If that's the case, I guess both that and the per-package-id option would require steps in the CI for uploading to some temporary remote after each profile/package-id build, so that steps further down the line could use the already built packages. For my personal case, that would be a thing that I'm currently missing. 😄

**Comment by solvingj** [comment]
@aleksa2808 
Many people struggle with this (in all languages in fact).  It's a fundamental challenge for any workflow scenarios where dependencies are obtained via package manager, and people need to test changes to multiple packages at a time (including other programming languages).

 There are currently two ways to address this with Conan:
1. For each commit/pull request, use the Artifactory API to provision a temporary Conan repository. Have the future steps use that as the upload location, and primary download location, falling back to the "production" repository for other packages.  At the end of successful pipelines, you can promote the binaries from the "temporary repo", to the "production repo".  It seems a bit "heavy-handed", but it does achieve the goal. Furthermore, it's actually the strategy that is used for Conan Center PRs, so we can say that it works fine with Artifactory at a non-trivial scale.

2.  Use Conan Lockfiles + Revisions, and a single permanent secondary Conan repository such as "conan-temp" which is only used by CI (not users).  This is one of the most important challenges lockfiles address.  For each commit or PR,  you start with a lockfile (there some options for this step).  But, once you have a lockfile, you can rebuild the package you've changed and the resulting lockfile will have the new revision.  You then upload that revision to the `conan-temp` repository. As long as all jobs are using lockfiles, the new revision won't affect any other unrelated jobs.  Now, after a rebuild of a library, you will typically need to build downstream consumers of the changed packaged as well, and so you pass the updated lockfile to those jobs.  This lockfile will bring in the specific revision you want from the upstream CI job, and will also update the lockfile with the revision of each downstream package which gets rebuilt as a result.  

With lockfiles and a single "conan-temp" repository, you can get completely isolated testing of a commit or PR, including the downstream rebuilds and tests.  However, be warned that lockfiles are still experimental, and still changing a lot.  Lockfiles also bring a number of other implications on the overall workflow, which take time to plan.  It is not a drop-in feature, so if you want to change the minimal amount, consider option 1 for now, and perhaps start research and experimentation with lockfiles now.

**Comment by memsharded** [comment]
The default package_id model for 2.0 is already in place, lets summarize:

- The environment is not part of the package_id. If it is something that fundamentally affects binary compatibility, then it should be modeled into settings or ``conf``.
- The typical use case of environment would be for example to define per machine variability, like the path of a compiler, that can be different in different machines, still exactly the same compiler
- Header-only libraries should not include anything in their package_id, it should always be exactly the same one. Except if for some reason there are headers that are generated differently with different inputs. That should be made most likely a setting or an option of the package.
